### DISCUSSION TITLE
Make JReleaser support natively maven (gradle) model

### DISCUSSION THREAD

**Discussion by rmannibucau** [jreleaser/jreleaser#104] [Ideas]
Listing deliverables is very verbose, error prone and troublesome when you have > 1 module.
Goal here is to leverage the build model (maven/gradle even if I'm clearly more a maven user ;)) to not have to list them explicitly or at least not them all but their "pattern" instead.

Pointers can be:

1. MavenProject.attachedArtifacts for most of them
2. Integrating with some well known plugins (out of my head assembly plugin when execution configuration is "detached" so missed by 1, jib-maven-plugin for docker images, etc...)

I think the configuration should be a list of included patterns (glob, regex or whatever which does the work is fine).
Default can include tar.gz, zip, binaries - classifier=arthur or starts with "linux" or type=bin to give a few examples of nice defaults) and docker images - guess mainly by reference and/or using jib properties files - with arthur plugin it is set in maven properties under arthur.image.imageId by default so guess filtering properties ending with "image.imageId" is an option. 

Then the configuration trick is about enabling to include or not asc, pom etc but it is mainly about filtering attached artifacts so shouldn't be crazy.

**Comment by aalmiray** [comment]
Thanks. A few things needed to be unpacked here:

 * JReleaser works with files in two ways:
   1. as a set of possibly unrelated files, added using the `files:` section.
   2. as a set of artifacts belonging to a distribution.
   The difference strives in that the first is just a set of files that can be checksumed, signed and uploaded to a release, while the latter can additionally be packaged and uploaded with Homebrew, Snap, etc.
* Files defined in the `files:` section can be specified either using an explicit path (with an `artifact:` block), or using a glob. https://jreleaser.org/guide/latest/configuration/files.html
* The `jreleaser-maven-plugin` should typically be defined in an aggregating POM, though it *may* be defined in more than one POM however the `release.update` property must be set to `true`, otherwise the build would fail.

If the `jreleaser-maven-plugin` were to grab files from `MavenProject.attachedArtifacts`, how should those files be categorized? At first, all of them could be added automatically to the `files:` section but that has the potential of uploading to a Git release all artifacts that are typically uploaded to a remote Maven repository (.pom, .jar, -sources.jar, -javadoc.jar, site.xml, etc). Some sort of filtering would be required. 

Now, for those artifacts that could be part of a distribution, how would JReleaser be able to tell the difference? Say the `shade` plugin is applied and the shaded artifact uses a explicit classifier (`-all`) so that both original and shaded artifacts are attached to the session. JReleaser would have to  know that the `-all` artifact should be added to a distribution of type `SINGLE_JAR` and ignore the other JAR artifact. Now, what if there's no `-all` classifier because the shaded artifact uses the default conventions, clobbering the original artifact?

How should artifacts that could possible belong to other distributions be handled? For example those created by the `jlink-maven-plugin` could be added to a distribution of type JLINK, but so do those created by the `helidon-maven-plugin` as it also creates jlink images. There are likely other plugins that produce jlink images as well. Same goes for Graal Native images. Micronaut, Helidon, Quarkus, and SpringBoot plugins can produce that kind of binaries that could be configured with a distribution of type `NATIVE_IMAGE`.

Grabbing artifacts from `MavenProject.attachedArtifacts` sounds like a good idea except that conditional filtering is required to organize them into `files:` and `distributions:`, as well as excluding some artifacts altogether.
OTOH integration with well known plugins will always be a catch-up game.

Here's a crazy idea given that I know little of the subject, what if JReleaser were to provide a Maven extension that becomes active for all projects in the session. This extension would function as a place to collect data related to artifacts you'd like to consume with JReleaser. A specific mojo would be used on each POM to mark artifacts and collect data. Then, similarly as it's done with the `nexus-staging-maven-plugin` (and other aggregating plugins), the last project executed in the session processes all the data and performs the appropriate operations.

**Reply by rmannibucau** [reply]
It is why I spoke about regex, this solves the categorization part solving the configuration verbosity and error prone point.

**Reply by aalmiray** [reply]
Alright, let's assume JReleaser can inspect `MavenProject.attachedArtifacts` for now, how would you configure the plugin to take advantage of this new capability? Examples include

1. single project build that produces a .zip file (appassembler/assembly). The zip file should be part of a `JAVA_BINARY` distribution.
  - requires discarding all attached artifacts except for the zip file.
  - must define matching artifacts as belonging to a `JAVA_BINARY` distribution _or_ potentially using .zip, .tar, tgz, .tar.gz as markers and automatically identify the distribution as `JAVA_BINARY`. Caveat `JLINK` distributions also support .zip files.

2. single project (Helidon/Micronaut/Quarkus) that produces a native image. The binary should be part of a `NATIVE_IMAGE` distribution.
  - requires identifying the native image binary which might not be attached to the session (fwk plugin impl dependent).
  - must define matching artifacts as belonging to a `NATIVE_IMAGE` distribution

3. multi project build that produces a shaded JAR (say 3 Maven modules: root, core, tool, where tool is the one that creates the shaded JAR). The shaded JAR does not use a custom classifier. Only the shaded JAR should be configured in a `SINGLE_JAR` distribution.
  - requires discarding all artifacts except for the shaded JAR that has shares the same filename layout as the non-shaded JAR from the sibling project (${project.artifactId}-${project.version}.jar)
  - given that the matching artifact ends with .jar the plugin could automatically configure the distribution type as `SINGLE_JAR`.

4. multi project build that produces a Jlink image packaged as a zip file. Only the zip file should be part of a `JLINK` distribution.
  - requires discarding all attached artifacts except for the zip file.
  - must define matching artifacts as belonging to a `JLINK` distribution, specially if `JAVA_BINARY` takes .zip as default.
5. a combination of 1, 3, and 4 (JReleaser itself has this combination)
  - requires matching each artifact and placing them in appropriate distribution types.

**Reply by rmannibucau** [reply]
1. with the regex, something along ${project.groupId}.assembly:*:*:linux can be an option but also enables "or" (`(linux.*|win.*)`).
2. for projects not enriching maven properties with the image ref you have to integrate with them reading their configuration from maven model (junit-platform-maven-plugin does it to migrate surefire configuration for ex)
3. 1 solves it
4. 1 solves it too
5. guess it solves it too if you can associate to the regex the distro type which is what i was envisioning

**Reply by aalmiray** [reply]
OK. If you were to configure this feature in a `pom.xml`, how would it look like?

**Reply by rmannibucau** [reply]
I had in mind a simple `List<DistributionSpec>` which is a `Distribution` but java becomes a regex and artifacts is no more there (can surely be a bit more simplified but high level it was the idea).

**Reply by aalmiray** [reply]
I see. It can get tricky. 

When `distribution.java` is not provided then `project.java` is used however that might not be what you want all the time. Example: JReleaser sets `project.java` to `8` because it's `JAVA_BINARY` distribution requires Java 8, but its `JLINK` and `SINGLE_JAR` distributions require Java 11 so they set their `distribution.java` to `11. It could had been done the other way round, project at `11` and the `SINGLE_JAR` at `8` however there would still be the need to make a difference. Thus I think keeping `distribution.java` is a must.

Using `regex` instead of `artifacts` could work but you'd be missing some additional metadata that certain distributions require, for example, all `JLINK` artifacts require a specific `platform`. The current platform _could_ be used if there is only a single matching artifact and the producing JDK is the same as the running JDK, but when that's not the case you need explicit configuration. You'd also miss the `transform:` property.

I guess that instead of removing  `artifacts` altogether, a `regex:` property could be added to the `Artifact` type, in such a way that you either define `path:` or `regex:`.  All matching artifacts would share the same `transform:` and `platform:`.

**Reply by rmannibucau** [reply]
Can work but configuration starts to look ambiguous.
Isnt platform "guessable" from classifier or artifact in all cases when it is needed?
If so no need to explicit it - it is what i expected at least.
Goal there is really to trim down and simplify the config, if it becomes too complicated it would be easier to do an extension configuring on the fly jreleaser (but hope we can avoid it since it is close to not use it - my current option for some big personal projects "prejreleaser" where an extension grabs assemblies and docker images based on the project conventions).

**Reply by aalmiray** [reply]
Yes, I suppose that the value of `platform:` could be matched against `artifact.classifier` when the latter exists.

So. We've covered the _what_ and the _how_, I'm still confused about the _when_, particularly on a multi-module build, as artifacts must be attached to the session before the final JReleaser configuration is evaluated and executed.

Say there's a project with the following structure

```root
├── pom.xml
├── project1
│   └── pom.xml
└── project2
    └── pom.xml
```

With `root` being both an aggregator (i.e, defines a `<modules>` section) and a parent POM for `project1` and `project2`. This means `root` is executed before the children modules thus the JReleaser configuration won't be ready if placed at that pom.xml. It can't be placed on `project1/pom.xml` either as that will make it miss the artifacts from `project2`. It could be placed at `project2` and that'll probably be fine. Til someone adds `project3` and then the JReleaser configuration must be moved from `project2` to `project3`.

What if there was a `release` module that takes care of all that? Such as

```root
├── pom.xml
├── project1
│   └── pom.xml
├── project2
│   └── pom.xml
└── release
    └── pom.xml
```

In this way the release settings would remain mostly unchanged as modules are added/removed form the build as needed. This would work as long as `release` is always the last module evaluated in the build. So far so good I guess. But then I've got another question: what happens when I'd like to perform a Maven release followed by a JReleaser release? All modules _except_ for `release` should be published to Maven Central. The `release` module is exclusive to building thus it should not be part of the published artifacts. And of course the `release` module must be invoked as part of the build lifecycle otherwise `MavenProject.attachedArtifacts` (or more accurately the Maven session) will not contain the required artifacts.

**Reply by rmannibucau** [reply]
Hmm, I don't expect jreleaser to do anything in all module until last one is reached as maven deploy plugin deployAtEnd otherwise it leads to a ton of issues and corruptions when the full build fails so guess when is "when the release fully suceeded".
No need to add a release module which is something we tend to avoid because it makes builds less maintainable in time.

**Reply by aalmiray** [reply]
That's the thing, I've never written a Maven plugin that exhibits behavior as Maven's deploy with `deployAtEnd`. Not only does JReleaser would have to be invoked _after_ deploy but also after the project that performed the deploy (with `deplotAtEnd` enabled when the build has more than 1 module). Any tips would be greatly appreciated as these are uncharted waters for me.

**Reply by rmannibucau** [reply]
idea is to use a static storage (or a session storage is my preference) and ensure the plugin is defined only once for the project (actually the key in maven plugin cache is unique which means gav is unique but more important there is no difference of repository between module but this is generally trivial to do).
Then you do something like https://github.com/apache/maven-deploy-plugin/blob/master/src/main/java/org/apache/maven/plugins/deploy/DeployMojo.java#L185 - note that you can also use a sisu/IoC singleton.
Being ran after deploy is as easy as defining the plugin after deploy plugin but binding it at the same phase (or a phase after if preferred).
Not sure that text block makes a lot of sense but tried to not be too verbose, let me know if you need a bit more or even chat on slack or hangout.

**Reply by aalmiray** [reply]
Alright, gave it more thought and here are some ideas:
- to be consistent with the other modes of usage, this mode must be explicitly activated, say by adding `<autoConfig>true<autoConfig>`.
- attached artifacts will only be used to configure distributions. If a developer desires to have artifacts be part of the generic `files` block then they must explicitly list them or use the glob option.
- for a given set of artifacts (those that have the same groupId and base artifactId) the default name of the distribution will be the artifactId, unless a explicit mapping is defined. The mapping key is artifactId.
- classifiers will be inspected to determine if a platform is needed (mac, windows, linux, etc).
- artifacts of type Zip/Tar default to distributions of type JAVA_BINARY unless explicitly configured (could be JLINK for example).
- artifacts of type Jar could be matched to a SINGLE_JAR distribution however more information is required to determine if it's a real überjar or just a plain jar, thus a special settings must match the distributionName. If an artifact of type Jar is found and no marking is in place then an error will occur.
- In order to avoid errors all over the place due to the previous rule, a list of includes and excludes may be defined. these lists define the artifactIds that may be processed as distributions.
- explicit `<distribution>` may be declared to define additional settings per distribution. Their name must match the mapped distribution name.

**Reply by rmannibucau** [reply]
sounds like a good start, quick coming wish can be to dryrun it to check it is what was expected but guess it is somehow already there, just needs to be pointed out in the doc probably next the autoconfig flag.

**Reply by aalmiray** [reply]
Ah sure, you can always run `jreleaser:config` to check the resolved configuration. Also, JReleaser generates a full trace log at `target/jreleaser/trace.log` on every invocation. This trace log contains the fully resolved model as well.

**Reply by aalmiray** [reply]
Hmm I may have spoken too soon. Invoking `jreleaser:config` to look at the resolved configuration would only make sense *if* it's bound to the `deploy` phase and it's invoked after all artifacts have been attached. The other option is to bind `jreleaser:release` or `jrelease:full-release` to the `deploy` phase, run it with `-Djreleaser.dryrun=true` set, and inspect the trace.log. The disadvantage of this is running a full deploy just to have a look at the resolved configuration. You might need to disable/skip the deploy plugin to avoid any mishaps.

**Reply by rmannibucau** [reply]
@aalmiray what if jreleaser becomes a maven extension and disables itself deploy plugin (both maven-deploy and oss sonatype one if exists)?
+1 for this dryrun flag

**Reply by aalmiray** [reply]
Yeah, that could work as long as the value of `releaser.dryrun` is readily available early on. First time crafting a Maven extension.

**Reply by rmannibucau** [reply]
Basically enables to rewrite completely the pom (don't abuse of it, model changes are not reflected into consumable pom but for plugins it is fine ;)). So dropping sonatype plugin if there and forcing deploy to have skip=true is just a matter of modifying the <configuration> (Xpp3Configuration) of the plugin :).

**Reply by aalmiray** [reply]
Here's another thought, given that it makes sense for the jreleaser goal (I suppose it must be `full-release`) to be always bound to the `deploy` phase I would  also like to have a version of `full-release` that's unbound. This would likely mean keeping `full-release` as it is right now (unbound) and adding another named goal that's bound. The added benefit is that `<autoConfig>` is no longer needed as this goal must have to work in that way.

**Reply by rmannibucau** [reply]
Maybe a sibling thought on that: never do anything at deploy time except tracking what is created (target/artifacts.json in the idea) and get jreleaser:post-release do all the jreleaser annouce work. It is a pain when a release fails because a cloud (external) service is down "right now".

**Comment by aalmiray** [comment]
One more thing, consuming artifacts of type "Docker image" might not make much sense, as if the image is already created by an external plugin the only thing left is publishing to a Docker registry. JReleaser can create a Dockerfile based on templates, a docker image based on said Dockerfile, and publish the image(s), but it does not accept (currently) a set of externally built images for publication as that may be a) done manually, or b) it's expected that the external image creator has the means for publishing images on its own.

Here I'd bring the popular microservice frameworks again (Micronaut, Helido, Quarkus, SpringBoot) as the already provide configuration for creating Docker images though I'm unsure if they also allow image publication as part of their toolchain. There are of course, other plugins that may be used to create docker image, as well as doing it by hand.

**Reply by rmannibucau** [reply]
Most of plugin wrapped jib to support image publication (spring, quarkus, arthur at least for the one i know well) in docker daemon or on a remote registry.
What jreleaser can do is to enable to use the image name/tag in messages (from group:artifact reference probably) and/or to push the related tar somewhere else (like github releases).
One common painpoint here is when the version is not the artifact version but a timestamped version so you must extract it from the build to be able to reference it in a message (twitter) or download/push it on a storage (github releases).
